api/proto/teleport/decision/v1alpha1/tls_identity.proto

```diff
@@ -0,0 +1,247 @@
+  repeated string database_users = 17;
+
+  // MFAVerified is the UUID of an MFA device when this Identity was
+  // confirmed immediately after an MFA check.
+  string mfa_verified = 18;
+
+  // PreviousIdentityExpires is the expiry time of the identity/cert that this
+  // identity/cert was derived from. It is used to determine a session's hard
+  // deadline in cases where both require_session_mfa and
+  // disconnect_expired_cert are enabled.
+  // See https://github.com/gravitational/teleport/issues/18544.
+  google.protobuf.Timestamp previous_identity_expires = 19;
+
+  // LoginIP is an observed IP of the client that this Identity represents.
+  string login_ip = 20;
+
+  // PinnedIP is an IP the certificate is pinned to.
+  string pinned_ip = 21;
+
+  // AWSRoleARNs is a list of allowed AWS role ARNs user can assume.
+  repeated string aws_role_arns = 22;
+
+  // AzureIdentities is a list of allowed Azure identities user can assume.
+  repeated string azure_identities = 23;
+
+  // GCPServiceAccounts is a list of allowed GCP service accounts that the user
+  // can assume.
+  repeated string gcp_service_accounts = 24;
+
+  // ActiveRequests is a list of UUIDs of active requests for this Identity.
+  repeated string active_requests = 25;
+
+  // DisallowReissue is a flag that, if set, instructs the auth server to deny
+  // any attempts to reissue new certificates while authenticated with this
+  // certificate.
+  bool disallow_reissue = 26;
+
+  // Renewable indicates that this identity is allowed to renew it's own
+  // credentials. This is only enabled for certificate renewal bots.
+  bool renewable = 27;
+
+  // Generation counts the number of times this certificate has been renewed.
+  uint64 generation = 28;
+
+  // BotName indicates the name of the Machine ID bot this identity was issued
+  // to, if any.
+  string bot_name = 29;
+
+  // BotInstanceID is a unique identifier for Machine ID bots that is persisted
+  // through renewals.
+  string bot_instance_id = 30;
+
+  // AllowedResourceIDs lists the resources the identity should be allowed to
+  // access.
+  repeated ResourceID allowed_resource_ids = 31;
+
+  // PrivateKeyPolicy is the private key policy supported by this identity.
+  string private_key_policy = 32;
+
+  // ConnectionDiagnosticID is used to add connection diagnostic messages when
+  // Testing a Connection.
+  string connection_diagnostic_id = 33;
+
+  // DeviceExtensions holds device-aware extensions for the identity.
+  DeviceExtensions device_extensions = 34;
+
+  // UserType indicates if the User was created by an SSO Provider or locally.
+  string user_type = 35;
+}
+
+// Trait encodes a single key-values pair from a user's traits.
+message Trait {
+  // name is the "key" of the trait.
+  string name = 1;
+  // values is the list of values the trait name maps to.
+  repeated string values = 2;
+}
+
+// RouteToApp holds routing information for applications.
+message RouteToApp {
```

Comment:
It'll be harder to get out of sync once we rely on the lib/decision packages for the actual authz logic.

Short-term it would be spotted once one of us hit that line. Bear in mind that these are an initial draft, not the ready-to-go definitions for the service.

(Also I'm open to suggestions.)

Reply:
Preventing drift is an ongoing topic of discussion, but we'd like to ideally prevent it programmatically.  My current thinking is that we'll likely want to use something like [`gofakeit.Faker.Struct`](https://pkg.go.dev/github.com/brianvoe/gofakeit/v7#Faker.Struct) to generate a fully filled identity via reflect, then use `gocmp` to verify that all fields are preserved across protobuf serialization/deserialization.

Reply:
In the case where both sources are protos it shouldn't be hard to compare the proto descriptors either.